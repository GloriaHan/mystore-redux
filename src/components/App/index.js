import React, { useState } from "react";
import Routing from "../Routing";
export const CartContext = React.createContext();
export const InputContext = React.createContext();

export function App(props) {
  const [productsInCart, setProductsInCart] = useState([]);
  const [inputValue, setInputValue] = useState("");

  return ( 
    <InputContext.Provider value={{ inputValue, setInputValue }}>
      <CartContext.Provider value={{ productsInCart, setProductsInCart }}>
        <Routing />
      </CartContext.Provider>
      </InputContext.Provider>

  );
}

export default App;
