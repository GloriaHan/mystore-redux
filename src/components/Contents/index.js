import React from "react";
import { Outlet } from "react-router-dom";
import Sidebar from "../Sidebar";
import { Root } from "./Contents.style";

export default function Contents() {
  return (
    <Root>
      <Sidebar />
      <Outlet />
    </Root>
  );
}
