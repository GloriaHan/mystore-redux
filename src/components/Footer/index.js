import React from "react";
import { Root, Img, Paymethods, SocialMedia, ContactUs } from "./Footer.style";
import ThumbUpOffAltIcon from "@mui/icons-material/ThumbUpOffAlt";
import Button from "@mui/material/Button";
export default function Footer() {
  return (
    <Root>
      <Paymethods>
        <span>Payment methods: &nbsp;</span>
        <Img
          src="https://cdn-icons-png.flaticon.com/512/39/39196.png"
          alt="visa"
        ></Img>
        <Img
          src="https://cdn-icons-png.flaticon.com/512/349/349237.png"
          alt="JCB"
        ></Img>
        <Img
          src="https://cdn-icons-png.flaticon.com/512/349/349247.png"
          alt="paypal"
        ></Img>
        <Img
          src="https://cdn-icons-png.flaticon.com/128/493/493771.png"
          alt="AMEX"
        ></Img>
      </Paymethods>
      <SocialMedia>
        <span>Like us? &nbsp;</span>
        <ThumbUpOffAltIcon />
        &nbsp; &nbsp;
        <Img
          src="https://cdn-icons-png.flaticon.com/128/733/733547.png"
          alt="facebook"
        ></Img>
        <Img
          src="https://cdn-icons-png.flaticon.com/512/2111/2111463.png"
          alt="instagram"
        ></Img>
        <Img
          src="https://cdn-icons-png.flaticon.com/512/1409/1409937.png"
          alt="twitter"
        ></Img>
        <Img
          src="https://cdn-icons-png.flaticon.com/128/5968/5968812.png"
          alt="ticktok"
        ></Img>
      </SocialMedia>

      <ContactUs sx={{ "& button": { m: 1 } }}>
        <Button variant="contained" size="small">
          Contact Us
        </Button>
      </ContactUs>
    </Root>
  );
}
