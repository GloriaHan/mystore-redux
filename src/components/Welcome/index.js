import React from "react";
import { useNavigate } from "react-router-dom";
import { Root } from "./Welcome.style";
import Button from "@mui/material/Button";

export default function Welcome() {
  const navigate = useNavigate();

  return (
    <Root>
      <h1> Homepage</h1>
      <h2>WELCOME TO MYSTORE !</h2>
      <h3>It's time to treat yourself.</h3>
      <div>
        <Button
          onClick={() => {
            navigate(`/mystore/products`);
          }}
          variant="contained"
          size="large"
        >
          Shopping Now
        </Button>
      </div>
    </Root>
  );
}
