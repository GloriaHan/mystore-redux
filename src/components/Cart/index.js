import React, { useContext } from "react";
import { CartContext } from "../App/index";
import {
  Root,
  Th,
  Img,
  Title,
  Qty,
  Product,
  Price,
  Table,
  Tr,
  Td,
  Thead,
  Tbody,
  Header,
  Container,
  StyledButton,
} from "./Cart.style";
import InputLabel from "@mui/material/InputLabel";
import MenuItem from "@mui/material/MenuItem";
import FormControl from "@mui/material/FormControl";
import Select from "@mui/material/Select";
import Box from "@mui/joy/Box";
import Chip from "@mui/joy/Chip";
import ChipDelete from "@mui/joy/ChipDelete";

export default function Cart() {
  const { productsInCart, setProductsInCart } = useContext(CartContext);

  const productsQty =
    productsInCart.length > 0
      ? productsInCart.reduce((a, b) => a + b.qty, 0)
      : 0;

  const totalPrice =
    productsInCart.length > 0
      ? productsInCart.reduce((a, b) => a + b.qty * b.price, 0)
      : 0;

  const addToCart = (e, item) => {
    let result = productsInCart.find((itemObj) => itemObj.id === item.id);
    if (result) {
      result.qty = e.target.value;
      setProductsInCart([...productsInCart]);
    }
  };
  const deleteProduct = (e, item) => {
    const newProductsList = [...productsInCart].filter((itemObj) => {
      return itemObj.id !== item.id;
    });
    setProductsInCart(newProductsList);
  };
  return (
    <Root>
      <Header>Your shopping Cart : {productsQty} items </Header>
      <Container>
        <Table>
          <Thead>
            <Tr>
              <Th>Product</Th>
              <Th>Quantity</Th>
              <Th>Unit Price</Th>
              <Th>Sub-Total</Th>
              <Th></Th>
            </Tr>
          </Thead>
          <Tbody>
            {productsInCart.map((item) => (
              <Tr key={item.id}>
                <Td>
                  <Product>
                    <Img src={item.image} alt={item.title} />
                    <Title>{item.title}</Title>
                  </Product>
                </Td>
                <Td>
                  <Qty>
                    <FormControl sx={{ m: 1, minWidth: 120 }} size="small">
                      <InputLabel id="demo-select-small">Qty</InputLabel>
                      <Select
                        labelId="demo-select-small"
                        id="demo-select-small"
                        value={item.qty}
                        label="Qty"
                        onChange={(e) => addToCart(e, item)}
                      >
                        {new Array(20).fill(1).map((item, index) => (
                          <MenuItem key={index} value={index + 1}>
                            {index + 1}
                          </MenuItem>
                        ))}
                      </Select>
                    </FormControl>
                  </Qty>
                </Td>
                <Td>
                  <Price>${item.price}</Price>
                </Td>
                <Td>
                  <Price>
                    {`$${Number(item.qty * item.price).toFixed(2)}`}
                  </Price>
                </Td>
                <Td>
                  <Box sx={{ display: "flex", gap: 1, alignItems: "center" }}>
                    <Chip
                      onClick={(e) => deleteProduct(e, item)}
                      variant="outlined"
                      color="danger"
                      endDecorator={<ChipDelete />}
                    >
                      Delete
                    </Chip>
                  </Box>
                </Td>
              </Tr>
            ))}
          </Tbody>
        </Table>
      </Container>

      <Container className="checkout">
        <div>
          <StyledButton variant="contained">Checkout</StyledButton>
          <p>items: {productsQty}</p>
          <p>Post fee : Free</p>
        </div>
        <div>
          <p>Total: ${`$${Number(totalPrice).toFixed(2)}`}</p>
        </div>
      </Container>
    </Root>
  );
}
