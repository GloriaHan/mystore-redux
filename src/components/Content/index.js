import React, { useEffect, useState, useContext } from "react";
import { useParams, useSearchParams } from "react-router-dom";
import { Root, Img, Title, Product, Price,NoResult } from "./Content.style";
import { useNavigate } from "react-router-dom";
import Box from "@mui/material/Box";
import Paper from "@mui/material/Paper";
import Button from "@mui/material/Button";
import { CartContext, InputContext } from "../App/index";

export default function Content() {
  const [loading, setLoading] = useState(false);
  const [products, setProducts] = useState([]);
  const { category } = useParams();
  const navigate = useNavigate();
  const [searchParams] = useSearchParams();
  const search = searchParams.get("search");
  const { productsInCart, setProductsInCart } = useContext(CartContext);
  const { setInputValue } = useContext(InputContext);

  useEffect(() => {
    (async () => {
      const url = `${process.env.REACT_APP_BASE_URL}/products${
        category ? `/category/${category}` : ""
      }`;
      setLoading(true);
      const res = await fetch(url);
      const data = await res.json();
      setLoading(false);
      const filterData = search
        ? data.filter(
            (item) =>
              item.title.toUpperCase().includes(search.toUpperCase()) ||
              item.description.toUpperCase().includes(search.toUpperCase())
          )
        : data;
      setProducts(filterData);
    })();
  }, [category, search]);

  if (loading === true) return null;
  return (
    <Root>
      {products.length <= 0 ? (
        <NoResult>
          <h3>Sorry, nothing matched your search.</h3>
          <h3>Please search again.</h3>
        </NoResult>
      ) : (
        products &&
        products.map((item) => (
          <Box
            key={item.id}
            sx={{
              display: "flex",
              flexWrap: "wrap",
              "& > :not(style)": {
                m: 1,
                width: 195,
                height: 260,
              },
            }}
          >
            <Paper elevation={2}>
              <Product
                onClick={() => {
                  navigate(`/products/${item.id}`);
                  setInputValue("");
                }}
              >
                <Img src={item.image} alt={item.title} />
                <Title>{item.title}</Title>
                <Price>${item.price}</Price>
                <Button
                  variant="contained"
                  size="small"
                  onClick={(e) => {
                    e.stopPropagation();
                    navigate("/cart");
                    setInputValue("");
                    let result = productsInCart.find(
                      (product) => product.id === item.id
                    );
                    if (result) {
                      result.qty = result.qty + 1 * 1;
                      setProductsInCart([...productsInCart]);
                    } else {
                      setProductsInCart([
                        ...productsInCart,
                        { ...item, qty: 1 },
                      ]);
                    }
                  }}
                >
                  BUY NOW
                </Button>
              </Product>
            </Paper>
          </Box>
        ))
      )}
    </Root>
  );
}
