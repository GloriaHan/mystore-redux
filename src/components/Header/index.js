import React, { useContext } from "react";
import { useNavigate } from "react-router-dom";
import {
  Img,
  Root,
  SearchBar,
  StyledButton,
  ShoppingCart,
  CartButton,
} from "./Header.style";
import shoppinglogo from "../../assets/image/shoppinglogo.png";
import Paper from "@mui/material/Paper";
import InputBase from "@mui/material/InputBase";
import IconButton from "@mui/material/IconButton";
import SearchIcon from "@mui/icons-material/Search";
import Badge from "@mui/material/Badge";
import Box from "@mui/material/Box";
import { CartContext, InputContext } from "../App/index";

export default function Header() {
  const { productsInCart } = useContext(CartContext);
  const { inputValue, setInputValue } = useContext(InputContext);
  const navigate = useNavigate();
  const productsQty =
    productsInCart.length > 0
      ? productsInCart.reduce((a, b) => a + b.qty, 0)
      : 0;

  return (
    <Root>
      <div  onClick={() => {
              navigate("/");
              setInputValue("");
            }}>
        <Img src={shoppinglogo} alt="logo" />
      </div>
      <SearchBar>
        <Paper
          component="form"
          sx={{
            p: "2px 4px",
            display: "flex",
            alignItems: "center",
            width: 400,
          }}
        >
          <InputBase
            sx={{ ml: 1, flex: 1 }}
            placeholder="Buying makes us happy"
            inputProps={{ "aria-label": "Buying makes us happy" }}
            type="text"
            value={inputValue}
            onChange={(e) => setInputValue(e.target.value)}
          />
          <IconButton sx={{ p: "10px" }} aria-label="search">
            <SearchIcon />
          </IconButton>
        </Paper>

        <Box sx={{ "& button": { m: 1 } }}>
          <StyledButton
            onClick={() => navigate(`/mystore/products?search=${inputValue}`)}
            variant="contained"
            size="large"
          >
            search
          </StyledButton>
        </Box>
      </SearchBar>

      <CartButton  onClick={() => {
              navigate("/cart");
              setInputValue("");
            }}>
        <Badge badgeContent={productsQty} color="error">
          <ShoppingCart fontSize="large" />
        </Badge>
      </CartButton>
    </Root>
  );
}
