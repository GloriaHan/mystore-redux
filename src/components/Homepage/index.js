import React from "react";
import { Outlet } from "react-router-dom";
import Header from "../Header";
import Footer from "../Footer";
import { Root, Content } from "./Homepage.style";

export default function Homepage() {
  return (
    <Root>
      <Header />
      <Content>
        <Outlet />
      </Content>
      <Footer />
    </Root>
  );
}
