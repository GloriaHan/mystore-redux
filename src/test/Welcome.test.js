import React from "react";
import "jest-styled-components";
import { BrowserRouter } from "react-router-dom";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { InputContext, CartContext } from "../components/App";
import Welcome from "../components/Welcome";

const productsInCart = [];

const mockedUsedNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedUsedNavigate,
}));

describe("Render Test", function () {
  it("case: expect rendering correct", () => {
    const { container } = render(
      <InputContext.Provider value={{}}>
        <CartContext.Provider value={{ productsInCart }}>
          <BrowserRouter>
            <Welcome />
          </BrowserRouter>
        </CartContext.Provider>
      </InputContext.Provider>
    );
    expect(container).toMatchSnapshot();
  });

  it("case: click button to trigger useNavigate", () => {
    render(
      <InputContext.Provider value={{}}>
        <CartContext.Provider value={{ productsInCart }}>
          <BrowserRouter>
            <Welcome />
          </BrowserRouter>
        </CartContext.Provider>
      </InputContext.Provider>
    );

    fireEvent.click(screen.getByText("Shopping Now"));
    expect(mockedUsedNavigate).toHaveBeenCalledTimes(1);
  });
});
