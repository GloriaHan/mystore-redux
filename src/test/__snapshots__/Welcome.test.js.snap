// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Test case: expect rendering correct 1`] = `
.c0 {
  text-align: center;
  background-image: url("https://cdn.vectorstock.com/i/1000x1000/80/01/digital-shop-logo-design-template-vector-14438001.webp");
  background-size: cover;
  color: #455a64;
  width: 100%;
}

<div>
  <div
    class="c0"
  >
    <h1>
       Homepage
    </h1>
    <h2>
      WELCOME TO MYSTORE !
    </h2>
    <h3>
      It's time to treat yourself.
    </h3>
    <div>
      <button
        class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButtonBase-root  css-jh47zj-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Shopping Now
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
